doc/tutorial.rst

```diff
@@ -0,0 +1,388 @@
+  $ pytest -v --backend ledgercomm test_something.py
+
+.. warning::
+  This won't work - again - because the tests will run for every SDK versions, but
+  the application currently started could only be of one SDK version!
+
+We can allow ``Pytest`` to specify the firmware by expanding/modifying our current code:
+
+.. code-block:: python
+  :caption: conftest.py
+  :linenos:
+
+  from pathlib import Path
+  from pytest import fixture
+
+  from ragger.firmware import Firmware
+  from ragger.backend import SpeculosBackend, LedgerCommBackend, LedgerWalletBackend, BackendInterface
+  from ragger.utils import app_path_from_app_name
+
+  APPS_DIRECTORY = Path("path/to/")
+  APP_NAME = "app"
+
+  FIRMWARES = [Firmware('nanos', '2.1'),
+               Firmware('nanox', '2.0.2'),
+              Firmware('nanosp', '1.0.3')]
+
+  BACKENDS = ["speculos", "ledgercomm", "ledgerwallet"]
+
+  def pytest_addoption(parser):
+      parser.addoption("--backend", action="store", default="speculos")
+      for fw in FIRMWARES:
+          parser.addoption("--"+fw.device, action="store_true", help=f"run on {fw.device} only")
+
+  @fixture(scope="session")
+  def backend_name(pytestconfig):
+      return pytestconfig.getoption("backend")
+
+  def create_backend(backend_name: str, firmware: Firmware, display: bool) -> BackendInterface:
+      if backend.lower() == "ledgercomm":
+          return LedgerCommBackend(firmware, interface="hid")
+      elif backend.lower() == "ledgerwallet":
+          return LedgerWalletBackend(firmware)
+      elif backend.lower() == "speculos":
+          args, kwargs = prepare_speculos_args(firmware, display)
+          return SpeculosBackend(*args, firmware, **kwargs)
+      else:
+          raise ValueError(f"Backend '{backend}' is unknown. Valid backends are: {BACKENDS}")
+
+  def pytest_generate_tests(metafunc):
+      if "firmware" in metafunc.fixturenames:
+          fw_list = []
+          ids = []
+          # First pass: enable only demanded firmwares
+          for fw in FIRMWARES:
+              if metafunc.config.getoption(fw.device):
+                  fw_list.append(fw)
+                  ids.append(fw.device + " " + fw.version)
+          # Second pass if no specific firmware demanded: add them all
+          if not fw_list:
+              for fw in FIRMWARES:
+                  fw_list.append(fw)
+                  ids.append(fw.device + " " + fw.version)
+          metafunc.parametrize("firmware", fw_list, ids=ids)
+
+  @fixture
+  def backend(firmware) -> SpeculosBackend:
+      app_location = app_path_from_app_name(APPS_DIRECTORY, APP_NAME, firmware.device)
+      backend = SpeculosBackend(app_location, firmware)
+      with backend:
+          yield backend
+
+And now we can easily specify which firmware should be used:
+
+.. code-block:: bash
+
+  $ # start the application on a physical device connected to the computer, or on an emulator
+  $ pytest -v --nanosp --backend ledgercomm test_something.py
```

Comment:
Il faut que le test_something prenne en paramètre le firmware pour qu'il soit fixturé non ? Peut être donner le nouveau template de test_something

Reply:
Hmm as the `backend` fixture is itself fixtured with `firmware`, this is enough no? Lemme check.

Reply:
Yep that's cool!